### PR TITLE
Update NodeJS to version 6

### LINKED ISSUE #536: Mise à jour Node à la version 6
En lien avec https://github.com/sgmap/mes-aides-api/issues/37

### PR DESCRIPTION
**Pull request by guillett**
Should fix #536 

Depends on https://github.com/sgmap/mes-aides-api/pull/113

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -7,7 +7,7 @@
     "angulartics-piwik": "^1.0.4",
     "bower": "^1.4.1",
     "cleave.js": "~0.7.15",
-    "errorhandler": "~1.3.0",
+    "errorhandler": "~1.5",
```
Comment (MattiSG) [inline package.json]: > [Development-only error handler middleware](https://github.com/expressjs/errorhandler).

…and indeed used only `if (app.get('env') == 'development')`. Should be moved to `devDependencies` :)

File: package.json [hunk 2]
```diff
@@ -22,49 +24,51 @@
...
+    "grunt-concurrent": "^2.3.0",
+    "grunt-contrib-clean": "^1.0.0",
+    "grunt-contrib-concat": "^1.0.1",
+    "grunt-contrib-copy": "^1.0.0",
+    "grunt-contrib-htmlmin": "^1.4.0",
+    "grunt-contrib-sass": "^1.0.0",
+    "grunt-contrib-uglify": "^1.0.1",
+    "grunt-contrib-watch": "^1.0.0",
+    "grunt-env": "^0.4.4",
+    "grunt-express-server": "*",
```
Comment (MattiSG) [inline package.json]: Are you sure you support **any** version of `grunt-express-server`?

File: Gruntfile.js [hunk 3]
```diff
@@ -155,15 +154,6 @@ module.exports = function (grunt) {
       }
     },
 
-    // Automatically inject Bower components into the app
-    'bower-install': {
```
Comment (MattiSG) [inline Gruntfile.js]: Why was this removed? The only information I have in the commit message is “Bump node to 6.9.5 and package versions” 😕

Has `bower-install` become redundant?

File: package.json [hunk 4]
```diff
@@ -22,49 +24,51 @@
...
+    "mocha": "^3.1.0",
+    "open": "0.0.5",
+    "phantomjs": "^2.1.7",
+    "requirejs": "^2.2.0",
+    "time-grunt": "^1.3.0",
     "watai": "^0.7.0"
   },
   "engines": {
-    "node": "0.10"
+    "node": ">=6.9.5"
```
Comment (MattiSG) [inline package.json]: Have you actually checked that you don't support Node v6.8? Be as lax as possible in the version constraints: once you've stated an engine version constraint, one has to trust you and getting the specific environment for running your app can be very costly.

File: package.json [hunk 5]
```diff
@@ -22,49 +24,51 @@
...
-    "open": "~0.0.4",
-    "time-grunt": "~1.4",
+    "karma-jasmine": "^1.0.2",
+    "karma-ng-html2js-preprocessor": "^1.0.0",
+    "karma-ng-scenario": "^1.0.0",
+    "karma-phantomjs-launcher": "^1.0.1",
+    "karma-requirejs": "^1.0.0",
+    "karma-script-launcher": "^1.0.0",
+    "load-grunt-config": "^0.19.2",
+    "mocha": "^3.1.0",
```
Comment (MattiSG) [inline package.json]: Why was the `mocha` version decreased?

File: package.json [hunk 6]
```diff
@@ -22,49 +24,51 @@
...
+    "open": "0.0.5",
+    "phantomjs": "^2.1.7",
+    "requirejs": "^2.2.0",
```
Comment (MattiSG) [inline package.json]: >A command line tool for running JavaScript scripts that use the Asynchronous Module Definition API (AMD) for declaring and using JavaScript modules and regular JavaScript script files.

I don't think we use AMD in the codebase. Why was this added as a dependency?

File: app/views/front.html [hunk 7]
```diff
@@ -110,7 +110,7 @@
     <script src="bower_components/angular-bootstrap/ui-bootstrap-tpls.js"></script>
     <script src="bower_components/ngstorage/ngStorage.js"></script>
     <script src="bower_components/moment/moment.js"></script>
-    <script src="bower_components/lodash/dist/lodash.compat.js"></script>
+    <script src="bower_components/lodash/dist/lodash.js"></script>
```
Comment (MattiSG) [inline app/views/front.html]: > This change would probably be a good opportunity to load `lodash` as an NPM dependency rather than a Bower dependency 😃

**Comment by MattiSG** [comment]
> - Remove supertest from devDependencies

Why? I can see you did that in the diff. Was it not used anywhere? Add this in the commit message:  `Remove unused 'supertest' dependency`.

> - Bump cleave version
> - Bump time-grunt version
> - Bump connect-livereload version
> - Bump errorhandler version

👍 
“Upgrade” would be more explicit than “Bump”. Adding the versions in the commit message can be useful as a way to see if there were major or minor updates.

> - Bump node to 6.9.5 and package versions

This commit is huge and should have been split in several ones.

> - to rm

Use `[TOREMOVE] Reason why the change was made`.

> - Bump lodash and obvious changes

Nothing is ever obvious to the external reader :wink:
“Upgrade lodash to v4” would have been enough, as updating API calls is definitely part of upgrading a dependency.

> - Trick lazy chaining to get filtered ressources

I didn't understand what happened there: what was the issue? Is it a consequence of some dependency update? If so, bundle that change as part of the update commit.

> - Update NodeJS version in CIs

👍

> - fix up bump to 6

Use the [`fixup!` syntax](http://fle.github.io/git-tip-keep-your-branch-clean-with-fixup-and-autosquash.html).

> - Update more dev dependencies
> - Update more dependencies
> - Update more dev dependencies

Nope. One commit per dependency is the best. If not, then group all those commits together, since “more” is only relative to the preceding commit.

**Comment by MattiSG** [comment]
Fails on my machine:

```
› npm install
[…]
sgmap-mes-aides-api@7.3.3 node_modules/sgmap-mes-aides-api
├── frozen-moment@0.4.0
├── codes-postaux@1.0.3
├── event-stream@3.3.4 (pause-stream@0.0.11, duplexer@0.1.1, stream-combiner@0.0.4, from@0.1.7, map-stream@0.1.0, split@0.3.3, through@2.3.8)
├── body-parser@1.17.1 (content-type@1.0.2, bytes@2.4.0, depd@1.1.0, qs@6.4.0, on-finished@2.3.0, raw-body@2.2.0, http-errors@1.6.1, iconv-lite@0.4.15, debug@2.6.1, type-is@1.6.15)
├── superagent@3.5.2 (cookiejar@2.1.1, component-emitter@1.2.1, methods@1.1.2, extend@3.0.0, mime@1.3.4, formidable@1.1.1, qs@6.4.0, debug@2.6.4, form-data@2.1.4, readable-stream@2.2.9)
├── underscore.string@3.3.4 (util-deprecate@1.0.2, sprintf-js@1.0.3)
├── moment@2.18.1
├── ludwig-api@1.2.1 (async@1.5.2, cookie-parser@1.4.3, passport-local@1.0.0, passport@0.3.2, bcryptjs@2.4.3, cors@2.8.3, express-session@1.15.2, connect-mongo@0.8.2, lodash@3.10.1)
└── kerberos@0.0.23 (nan@2.5.1)
› node --version
v7.7.4
› git rev-parse HEAD
9809f18116560a4b0f4521276df6e24784cf6f4a
› npm start
> sgmap-mes-aides-ui@1.0.0 prestart /Users/sgmap/Dev/mes-aides-ui
> grunt build

Loading "Gruntfile.js" tasks...ERROR
>> Error: Cannot find module 'load-grunt-tasks'
Warning: Task "build" not found. Use --force to continue.

Aborted due to warnings.
```

File: Gruntfile.js (hunk 3 above)
Comment (guillett) [inline Gruntfile.js]: Yes, it wasn't use and a direct call to bower is in `postinstall`.

**Comment by guillett** [comment]
It should work on your station. `load-grunt-config` is in package.json.

Do you have anything suspicious in your `npm install` logs?

**Comment by guillett** [comment]
- `load-grunt-config` depends on `load-grunt-tasks`
- `load-grunt-tasks`was not directly required in package.json

That's probably why it fails...

I am a bit confused as I used yarn and npm without that error and in CI and in Vagrant it works too...

**Comment by guillett** [comment]
- [ ] I am pretty sure I ran all Ludwig tests but I will run it before merging this PR to master.

**Comment by fpagnoux** [comment]
Seems to work.
